enhancements/security/scc-run-as-any.md

```diff
@@ -0,0 +1,134 @@
+This ordering is critical to correctly matching SCCs to Pods, and must be clear an consistent.
+
+The `RunAsAny` strategy is documented as being the _least restrictive_ `runAsUser` strategy.
+In other words, a Pod that runs under some other `runAsUser` strategy must also run under `RunAsAny`
+
+This is not currently the case for pods with `runAsNonRoot:true`. They run correctly under `MustRunAsRange` but fail under `RunAsAny`.
+If a Pod has no `runAsUser` and no image `User` setting then:
+- Under `MustRunAsRange`, SCC processing "helps" by providing a default non-root UID from a predefined range.
+- Under `RunAsAny`, SCC processing does not help, and assumes the Docker default of 0. The Pod is rejected by its own security context.
+
+The proposal: For `runAsNonRoot:true` pods, `RunAsAny` should use the same defaulting rules as `MustRunAsRange` to provide a non-0 UID.
+
+[scc]: https://docs.openshift.com/container-platform/4.12/authentication/managing-security-context-constraints.html
+
+## Motivation
+
+Major customer outrage, caused in this instance by the Cluster Logging Operator Pod failing.
+
+- [Upstart Case 03508387 Review](https://docs.google.com/document/d/1AlVZHsWqykMELnJmE7Ei-wYCDAPEA966zPZnVDnkHIQ/edit)
+- [OCPBUGS-13375 Logging Operator 5.7.0 fails with "CreateContainerConfigError" when there are certain custom SCCs in the cluster](https://issues.redhat.com/browse/OCPBUGS-13375)
+
+Note the operator has been fixed to resolve the customer problem, but that does not resolve the larger problem that adding a custom SCC can break a cluster.
+
+The operator does not express any preference for UID except that it not be root.
+It has been running correctly in its current configuration for ages because it matched a standard SCC with `MustRunAsRange`.
+
+`MustRunAsRange` does not simply constrain the UID, it _provides a valid default_ `runAsUser` field, if there isn't one otherwise specified.
+
+To be strictly "less restrictive" `RunAsAny`, also needs to provide a valid runAsUser by default. 0 is a valid and sensible default with `runAsNonRoot:false`, but it is INVALID with `runAsNonRoot:true`.
+
+The customer (correctly) assumed that a custom SCC with `RunAsAny` was safe because it is less restrictive.
+Any pods already working in their cluster should continue to work.
+
+This was not their experience.
+
+**Note 1** A `runAsNonRoot` Pod _should_ have an explicit `runAsUser` or set the `User` property in the image. That is not the point of this proposal.
+The point is that SCC processing must be consistent. We can't say "your pods are wrong" when a customer adds a _less restrictive_ SCC and Pods that used to work suddenly fail.
+The user cannot control the match between SCC and Pod (that is the whole point of an SCC), so its up to us to ensure that the definition of "less restrictive" rigorously matches Pod behaviour.
```

Comment:
I am not an expert but doen't that defeat the purpose of the SCC mechanism? Isn't the goal to allow the user to define _new_ SCCs that will automatically apply to _existing_ Pod? If  you allow a pod to pick an SCC, then you might as well throw away SCSs and just specify everything on the existing, well-defined Pod SecurityContext.

Reply:
> Isn't the goal to allow the user to define new SCCs that will automatically apply to existing Pod

No, the goal is to give admins a way to constrain what sort of permissions users can give to the pods they create.

And changing SCCs has no impact on already running pods (they were already admitted), though it might affect a redeploy(when the deployment recreates the pods for one reason or another).

But we already tell admins not to change the default SCCs and i'd expect this feature to only be used to specify a default SCC in most cases, so all you're saying is "i should always use this SCC which should always exist on the cluster and never be changed"

> If you allow a pod to pick an SCC, then you might as well throw away SCSs and just specify everything on the existing, well-defined Pod SecurityContext.

no, because you still need a way to enforce whether the user is *allowed* to specify certain values on the security context (e.g. whether they are allowed to specify privileged=true), and that's what SCC does (and now what upstream k8s is finally introducing with pod security admission, though they also tried earlier w/ pod security policy)

To be clear, introducing a way for a pod to say "i want to run under this SCC" doesn't mean the pod(user) is guaranteed to be allowed to use that SCC.  So the pod might be rejected as "sorry, you said you only want to run under foo SCC, but you don't have permission to use foo SCC, so either remove your request, or request a different SCC you have access to".  And *also* the pod's securitycontext would need to meet the requirements of the required SCC.

Reply:
> Allowing workloads to request a specific SCC would also be a useful feature and one that could have been used to solve this problem if it had been available. As I recall it's been discussed before and may even have a draft EP. @deads2k or @stlaz do you have a pointer?

Implementation PR here: https://github.com/openshift/apiserver-library-go/pull/90

Reply:
I think pinning to a specific SCC defeats the purpose of SCC matching, in that case it's just an awkward, indirect way of writing the SC. But maybe I am missing the point of SCCs entirely.

Reply:
The primary purpose of SCCs is so that a cluster administrator can decide what permissions users should be allowed to claim when creating pods.

Allowing a user to say "this is the SCC i want this pod to be constrained to, i don't want this pod to accidentally be granted any additional permissions even if i have access to those additional permissions" actually extends that ability to a user who can now constrain themselves/their workloads explicitly, by requesting a specific SCC.

So it has value for that reason if nothing else.

It also happens to be a "weird indirect way of writing the SC" (in the case where you request an SCC that enforces MustRunAsRange), but even in that case, it's solving a problem the user can't solve themselves, because they cannot a priori set the RunAsUser to a valid value on their pod definition.  (I think there are better ways to solve that problem, as we've discussed elsewhere, but it would be a way to solve it)

All the mutation behavior, is, imho, confusing and somewhat orthogonal behavior.  With RunAsRange mutation being a particular hack that was done to set a particular field of the SC to a value the user could not possibly supply themselves.

Reply:
+1 the mixture of constraint and mutation is exactly what makes the current semantics hard to understand and work with.

Reply:
thanks!